pkg/controller/usercleanup/user_cleanup_controller.go

```diff
@@ -0,0 +1,166 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	"time"
+)
+
+var log = logf.Log.WithName("controller_usercleanup")
+
+type StatusUpdater func(userAcc *toolchainv1alpha1.UserSignup, message string) error
+
+// Add creates a new UserCleanup Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, crtConfig *crtCfg.Config) error {
+	return add(mgr, newReconciler(mgr, crtConfig))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, crtConfig *crtCfg.Config) reconcile.Reconciler {
+	return &ReconcileUserCleanup{
+		client:    mgr.GetClient(),
+		scheme:    mgr.GetScheme(),
+		crtConfig: crtConfig,
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("usercleanup-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to the secondary resource UserSignup and requeue the owner UserSignup
+	if err := c.Watch(
+		&source.Kind{Type: &toolchainv1alpha1.UserSignup{}},
+		&handler.EnqueueRequestForOwner{
+			IsController: true,
+			OwnerType:    &toolchainv1alpha1.UserSignup{},
+		}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileUserCleanup{}
+
+// ReconcileUserCleanup cleans up old UserSignup resources
+type ReconcileUserCleanup struct {
+	client    client.Client
+	scheme    *runtime.Scheme
+	crtConfig *crtCfg.Config
+}
+
+// Reconcile reads that state of the cluster for a UserSignup object and makes changes based on the state read
+// and what is in the UserSignup.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileUserCleanup) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling UserSignup")
+
+	// Fetch the UserSignup instance
+	instance := &toolchainv1alpha1.UserSignup{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+	reqLogger = reqLogger.WithValues("username", instance.Spec.Username)
+
+	if instance.Spec.VerificationRequired && !instance.Spec.Approved {
```

Comment:
When could it happen that the `VerificationRequired` is true and at the same time the `Approved` is true as well?

Reply:
It probably can't if everything works as expected but I would keep that check just to be on the safe side. I don't want us accidentally delete an active signup.

Reply:
I included this check as insurance, just to be safe as Alexey mentions.